src/include/txn.h

```diff
@@ -99,13 +99,60 @@ typedef enum {
         txn_state->pinned_id = saved_state.pinned_id;                     \
     } while (0)
 
-struct __wt_txn_state {
+struct __wt_txn_shared {
     WT_CACHE_LINE_PAD_BEGIN
     volatile uint64_t id;
     volatile uint64_t pinned_id;
     volatile uint64_t metadata_pinned;
     volatile bool is_allocating;
 
+    /*
+     * Durable timestamp copied into updates created by this transaction. It is used to decide
+     * whether to consider this update to be persisted or not by stable checkpoint.
+     */
+    wt_timestamp_t durable_timestamp;
+
+    /*
+     * Set to the first commit timestamp used in the transaction and fixed while the transaction is
+     * on the public list of committed timestamps.
+     */
+    wt_timestamp_t first_commit_timestamp;
+
+    /*
+     * Set to the first read timestamp used in the transaction. As part of our history store
+     * mechanism, we can move the read timestamp forward so we need to keep track of the original
+     * read timestamp to know what history should be pinned in front of oldest.
+     */
+    wt_timestamp_t pinned_read_timestamp;
+
+    TAILQ_ENTRY(__wt_txn_shared) read_timestampq;
+    bool clear_read_q; /* Set if need to clear from the read queue */
+    TAILQ_ENTRY(__wt_txn_shared) durable_timestampq;
+    /* Set if need to clear from the durable queue */
+    bool clear_durable_q;
+
+    WT_TXN *internal_txn; /* A reference to the corresponding internal transaction data. */
```

Comment:
5f9ebaa

Removed slop.

As for the `internal_txn` member, yeah we usually shouldn't be looking at that from other threads. The only exception (and this is why I've left this escape hatch here) is in `__txn_get_published_timestamp`. We need to compare the `commit_timestamp` with `first_commit_timestamp` to figure out whether we should return the `first_commit_timestamp` or in the case of a potentially resolved prepared transaction, `durable_timestamp`. We specifically have to check the timestamps because the transaction may be committed before getting removed from the queue and may have its flags cleared.